python/pyspark/ml/image.py

```diff
@@ -0,0 +1,196 @@
+
+        .. versionadded:: 2.3.0
+        """
+
+        if self._undefinedImageType is None:
+            ctx = SparkContext._active_spark_context
+            self._undefinedImageType = \
+                ctx._jvm.org.apache.spark.ml.image.ImageSchema.undefinedImageType()
+        return self._undefinedImageType
+
+    def toNDArray(self, image):
+        """
+        Converts an image to a one-dimensional array.
+
+        :param image: The image to be converted
+        :rtype array: The image as a one-dimensional array
+
+        .. versionadded:: 2.3.0
+        """
+
+        height = image.height
+        width = image.width
+        nChannels = image.nChannels
+        return np.ndarray(
+            shape=(height, width, nChannels),
+            dtype=np.uint8,
+            buffer=image.data,
+            strides=(width * nChannels, nChannels, 1))
+
+    def toImage(self, array, origin=""):
+        """
+        Converts an array with metadata to a two-dimensional image.
+
+        :param array array: The array to convert to image
+        :param str origin: Path to the image, optional
+        :rtype object: Two dimensional image
+
+        .. versionadded:: 2.3.0
+        """
+
+        if array.ndim != 3:
+            raise ValueError("Invalid array shape")
+        height, width, nChannels = array.shape
+        ocvTypes = ImageSchema.ocvTypes
+        if nChannels == 1:
+            mode = ocvTypes["CV_8UC1"]
+        elif nChannels == 3:
+            mode = ocvTypes["CV_8UC3"]
+        elif nChannels == 4:
+            mode = ocvTypes["CV_8UC4"]
+        else:
+            raise ValueError("Invalid number of channels")
+        data = bytearray(array.astype(dtype=np.uint8).ravel())
+        # Creating new Row with _create_row(), because Row(name = value, ... )
+        # orders fields by name, which conflicts with expected schema order
+        # when the new DataFrame is created by UDF
+        return _create_row(self.imageFields,
+                           [origin, height, width, nChannels, mode, data])
+
+    def readImages(self, path, recursive=False, numPartitions=-1,
+                   dropImageFailures=False, sampleRatio=1.0, seed=0):
+        """
+        Reads the directory of images from the local or remote source.
+
+        WARNINGS:
+          - If multiple jobs are run in parallel with different sampleRatio or recursive flag,
+            there may be a race condition where one job overwrites the hadoop configs of another.
+
+        :param str path: Path to the image directory
+        :param bool recursive: Recursive search flag
+        :param int numPartitions: Number of DataFrame partitions
+        :param bool dropImageFailures: Drop the files that are not valid images
+        :param float sampleRatio: Fraction of the images loaded
+        :param int seed: Random number seed
+        :rtype DataFrame: DataFrame with a single column of "images",
```

Comment:
I am sorry @imatiach-msft for a forth-and-back. I was confused between my production code and spark codes. Looks we should better use `:return:` just for consistency with other docs.

Reply:
done